### DISCUSSION TITLE
Memory problem

### DISCUSSION THREAD

**Discussion by hglee** [leppie/IronScheme#140] [Discussions] [answered]
Is there any solution to a memory problem?

I know there are limitation for dynamic IL that could not be unloaded.

But is there any trick to avoid memory problem for long running program?

This is simple sample code just eval loop.

Snapshot 1 at 'Start' and Snapshot 2 at 'End'.

```
            // Snapshot #1
            Console.WriteLine("Start");
            Console.ReadLine();

            for (int i = 0; i < 1000; ++i)
            {
                IronScheme.RuntimeExtensions.Eval("(+ 1 2)");
            }

            Console.WriteLine("Before GC");
            Console.ReadLine();

            GC.Collect(2, GCCollectionMode.Forced);
            GC.WaitForPendingFinalizers();

            // Snapshot #2
            Console.WriteLine("End");
            Console.ReadLine();
```

![memprofile](https://user-images.githubusercontent.com/427595/200538595-f92fcb23-da19-430f-bb02-251bd3a57387.png)

**Comment by leppie** [comment, marked as answer]
The problem is that it is compiling 1000 programs (and I dont really know if that code code can be GC'd).

The correct way would to use `Eval` to only open the door for interop. 

Example:

```
var add = "+".Eval<Callable>(); // now we have direct reference to the + proc

for (int i = 0, i < 1000; i++)
{
  add.Call(1, 2); // args are boxed, but will be GC'd
}
```

The above code does not even call the compiler as it is just simple lookup to the `+` proc.

Also, on today's hardware, that extra 15MB is not much :D Might be on ARM systems though.

I would also much sure to call `Eval` at least once before taking the snapshot, so you can be sure the runtime has been loaded. Might find it is actually less than 15MB growth.

Lastly, `Eval` has a lot of overhead (to deal with environments and positional args) and it is not really suited for tight loops.

**Reply by hglee** [reply]
That could be a solution to fixed contents script with known entry points.

And it needs to prepare for long running program with unpredictable scripts. For example, interactive REPL embedded to GUI application.

Anyway, I attached profile data with more point. It takes almost 10 KBytes per Eval.

```
            // Snapshot #1
            Console.WriteLine("Start");
            Console.ReadLine();

            IronScheme.RuntimeExtensions.Eval("(= 1 1)");

            GC.Collect(2, GCCollectionMode.Forced);
            GC.WaitForPendingFinalizers();

            // Snapshot #2
            Console.ReadLine();

            for (int i = 0; i < 1000; ++i)
            {
                IronScheme.RuntimeExtensions.Eval("(+ 1 2)");
            }

            Console.WriteLine("Before GC");
            Console.ReadLine();

            GC.Collect(2, GCCollectionMode.Forced);
            GC.WaitForPendingFinalizers();

            // Snapshot #3
            Console.WriteLine("End");
            Console.ReadLine();
```
![prof](https://user-images.githubusercontent.com/427595/201235893-6e9d1fe5-07a6-4c14-b273-59f223382e43.PNG)

**Reply by leppie** [reply]
If it is really a strict requirement, you probably better off running a Scheme interpreter. There are many quality ones. 

IronScheme is an incremental compiler and the REPL is just a tool to aid in writing programs.
